File: java-api-core/test/com/kinvey/java/annotations/AnnotationProcessingTest.java
package com.kinvey.java.annotations;

import com.google.api.client.json.GenericJson;
import com.google.api.client.util.Key;
import java.util.List;

import junit.framework.TestCase;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Map;

/**
 * Created by Prots on 3/11/16.
 */
public class AnnotationProcessingTest extends TestCase {


    @Test
    public void testInlineObjects(){
        GenericJson genericJson = ReferenceHelper.processReferences(new GenericJson() {
            @Key("test")
            @KinveyReference(collection = "test", itemClass = SampleGson.class)
            GenericJson innerObject = new SampleGson("test");
        }, new ReferenceHelper.ReferenceListener() {
            @Override
            public String onUnsavedReferenceFound(String collection, GenericJson object) {
                return "test";
            }
        });

        assertNotNull(genericJson.get("test"));
        assertTrue(Map.class.isAssignableFrom(genericJson.getClass()));
        Map reference = (GenericJson)genericJson.get("test");

        assertNotNull(reference.get("_id"));
        assertEquals(reference.get("_id"), "test");
        assertEquals(reference.get("_collection"), "test");
    }


    @Test
    public void testInlineCollection(){

        final List<SampleGson> inner = new ArrayList<SampleGson>();
        for (int i = 0 ; i < 10; i++){
            inner.add(new SampleGson(String.valueOf(i)));
        }

        GenericJson genericJson = ReferenceHelper.processReferences(new GenericJson() {
            @Key("test")
            @KinveyReference(collection = "test", itemClass = SampleGson.class)
            List<SampleGson> innerObject = inner;
        }, new ReferenceHelper.ReferenceListener() {
            @Override
            public String onUnsavedReferenceFound(String collection, GenericJson object) {
                return "test";
            }
        });

        assertNotNull(genericJson.get("test"));
        assertTrue(Map.class.isAssignableFrom(genericJson.getClass()));
        List reference = (List)genericJson.get("test");
        assertEquals(reference.size(), 10);

    }

    public static class SampleGson extends GenericJson{
        @Key("data")
        String id;

        public SampleGson(String id) {
            this.id = id;
        }
    }

}

File: java-api-core/test/com/kinvey/java/annotations/AnnotationProcessingTest.kt
package com.kinvey.java.annotations

import com.google.api.client.json.GenericJson
import com.google.api.client.util.Key
import com.kinvey.java.annotations.ReferenceHelper.ReferenceListener
import com.kinvey.java.annotations.ReferenceHelper.processReferences
import junit.framework.TestCase
import org.junit.Test
import java.util.*

/**
 * Created by Prots on 3/11/16.
 */
class AnnotationProcessingTest : TestCase() {
    @Test
    @Throws(InstantiationException::class, IllegalAccessException::class)
    fun testInlineObjects() {
        val genericJson = processReferences(object : GenericJson() {
            @KinveyReference(fieldName = "test", collection = "test", itemClass = AnnotationProcessingTest.SampleGson::class)
            var innerObject = SampleGson("test")
        }, object : ReferenceListener {
            override fun onUnsavedReferenceFound(collection: String, item: GenericJson?): String {
                return "test"
            }
        })
        assertNotNull(genericJson!!["test"])
        assertTrue(Map::class.java.isAssignableFrom(genericJson.javaClass))
        val reference = genericJson["test"] as GenericJson
        assertNotNull(reference["_id"])
        assertEquals(reference["_id"], "test")
        assertEquals(reference["_collection"], "test")
    }

    @Test
    @Throws(InstantiationException::class, IllegalAccessException::class)
    fun testInlineCollection() {
        val inner = (0..9).mapNotNull { i -> SampleGson(i.toString()) }
        val genericJson = processReferences(object : GenericJson() {
            @KinveyReference(fieldName = "test", collection = "test", itemClass = AnnotationProcessingTest.SampleGson::class)
            var innerObject: List<SampleGson> = inner
        }, object : ReferenceListener {
            override fun onUnsavedReferenceFound(collection: String, item: GenericJson?): String {
                return "test"
            }
        })
        assertNotNull(genericJson!!["test"])
        assertTrue(Map::class.java.isAssignableFrom(genericJson.javaClass))
        val reference = genericJson["test"] as List<*>
        assertEquals(reference.size, 10)
    }

    class SampleGson(@field:Key("data") internal var id: String) : GenericJson()
}